tests/how_to_test_systems.rs

```diff
@@ -0,0 +1,113 @@
+use std::collections::HashMap;
+
+use bevy::{ecs::system::CommandQueue, prelude::*};
+
+struct Enemy {
+    hit_points: u32,
+}
+
+struct CharacterTemplate {
+    hit_points: HashMap<&'static str, u32>,
+}
+
+fn despawn_dead_enemies(mut commands: Commands, enemies: Query<(Entity, &Enemy)>) {
+    for (entity, enemy) in enemies.iter() {
+        if enemy.hit_points == 0 {
+            commands.entity(entity).despawn_recursive();
+        }
+    }
+}
+
+fn hurt_enemies(mut enemies: Query<&mut Enemy>) {
+    for mut enemy in enemies.iter_mut() {
+        enemy.hit_points -= 1;
+    }
+}
+
+fn spawn_enemy(mut commands: Commands, character_template: Res<CharacterTemplate>) {
+    commands.spawn().insert(Enemy {
+        hit_points: *character_template.hit_points.get("enemy").unwrap(),
+    });
+}
+
+#[test]
+fn did_hurt_enemy() {
+    // Setup world and commands
+    let mut world = World::default();
+    let mut queue = CommandQueue::default();
+    let mut commands = Commands::new(&mut queue, &world);
+
+    // Setup stage with our two systems
+    let mut update_stage = SystemStage::parallel();
+    update_stage.add_system(hurt_enemies.system().before("death"));
+    update_stage.add_system(despawn_dead_enemies.system().label("death"));
+    let mut schedule = Schedule::default();
+    schedule.add_stage("update", update_stage);
+
+    // Setup test entities
+    let ennemy_id = commands.spawn().insert(Enemy { hit_points: 5 }).id();
+    queue.apply(&mut world);
+
+    // Run systems
+    schedule.run(&mut world);
+
+    // Check resulting changes
+    assert!(world.get::<Enemy>(ennemy_id).is_some());
+    assert_eq!(world.get::<Enemy>(ennemy_id).unwrap().hit_points, 4);
+}
+
+#[test]
+fn did_despawn_enemy() {
+    // Setup world and commands
+    let mut world = World::default();
+    let mut queue = CommandQueue::default();
+    let mut commands = Commands::new(&mut queue, &world);
+
+    // Setup stage with our two systems
+    let mut update_stage = SystemStage::parallel();
+    update_stage.add_system(hurt_enemies.system().before("death"));
+    update_stage.add_system(despawn_dead_enemies.system().label("death"));
+    let mut schedule = Schedule::default();
+    schedule.add_stage("update", update_stage);
+
+    // Setup test entities
+    let ennemy_id = commands.spawn().insert(Enemy { hit_points: 1 }).id();
+    queue.apply(&mut world);
```

Comment:
For some variety, perhaps some of these examples should add their entities/resources to the world directly (with direct world access APIs), rather than setting up and using `Commands` in every example?

Reply:
This example looks like a copy-paste of the previous one, so maybe not here (if you want to keep them consistent to make it obvious). You could show direct world access in the next one, but up to you. I just think there should be more variety in the APIs used.

Reply:
Tough call. I think being consistent here helps us be more prescriptive. Examples are a form of "style recommendation" where we can help encourage our users to write code a certain way. If we do different things (for the sake of being different), I personally consider that to be a bit confusing (and gives up a chance to set a standard). The user might ask "what makes these two examples different" or "which style should I use for my tests".

I think using commands is probably the correct approach if we are spawning from a system. But if someone has direct world access, I see no reason to add the additional boilerplate of Commands. Imo we should train users to use World directly when they have the chance, both for performance and terseness.

Reply:
very good point, I'm actually not used to use world directly so this is a good case for me too!